base_layer/core/src/base_node/service/service.rs

```diff
@@ -407,13 +407,68 @@ async fn handle_incoming_request<B: BlockchainBackend + 'static>(
         response: Some(response.into()),
     };
 
-    outbound_message_service
+    trace!(
+        target: LOG_TARGET,
+        "Attempting outbound message in response to inbound request ({})",
+        inner_msg.request_key
+    );
+    match outbound_message_service
         .send_direct(
             origin_public_key,
             OutboundEncryption::None,
             OutboundDomainMessage::new(TariMessageType::BaseNodeResponse, message),
         )
-        .await?;
+        .await
+    {
```

Comment:
So adding this in does give more detail in terms of the send process but it also might have some unintended consequences.

The prior code started the send and then moved on. This added code adds the `result.resolve_ok().await` which will wait for discovery to complete if it needs to and `send_states.wait_single().await ` will wait for the send to pass or fail which can take time depending on the Tor situation.

So I just want to make sure that this potential extra time (up to 2 minutes for discovery and up to 6 minutes by default on the Tor connection deciding that the destination is unreachable) will not derail other tasks within the BN?

Reply:
The way I understand is that `if send_states.len() == 1` the next `send_states.wait_single().await` follows directly. We could hav a look at the timing in the logs, with and without send errors.

Reply:
Ultimately a call to `send_direct` calls `pub async fn send_raw`, which will wait for a reply message from the receiver over the wire. This is where all the significant waiting will happen if any. The handling of message states afterwards is internal and quick.

Reply:
So the first thing that can take long is that you have added the `result.resolve_ok().await`. If discovery has to happen then that await can wait for multiple minutes. If we don't need to do discovery then the message is queued to be sent.

Once the message is queued we wait at `send_states.wait_single().await`. This explicitly waits until the queued message is sent or fails to send. Usually this takes about 8 seconds BUT if the other party is not reachable we have seen tor take up to 6-8 minutes to decide that the other party cannot be reached.

With the code you are replacing all of the discovery and waiting for the queued message send is not waited on in this code path. It treats the send as a fire and forget and lets the comms stack handle it. By adding this code in you are explicitly waiting on the various stages of sending. So the question is: Can this code tolerate those long waits?

Reply:
Understood.. lets wait then

Reply:
All added log messages dependent on `send_states. ... .await` are now executed from within a `tokio::spawn(async move` task, so no blocking calls anymore.